### PR TITLE
Kahypar ordering

### PR DESCRIPTION
**Pull request by danlkv**

### CODE REVIEW

**Comment by danlkv** [comment]
@ZichangHe please use this branch to publish your code. Use ./qtensor/optimization/kahypar_ordering/generate_TN.py to generate a TN hypergraph for a circuit. See changes of this merge request. Put your code in the kahypar_ordering directory.

**Comment by danlkv** [comment]
Awesome! I run the tests and it works quite well. I like how you did the comparison between different algorithms. Some potential improvements/questions for code:

1. The partite list data structure now supports only binary trees. I think extending it to a more tree-like structure will help support 3+ part partitions and also will simplify searching for children in line 137
2. The function is a bit large and hard to comprehend. I would decompose it into sub-functions like `find_children(partition)` and `get_vertex_intersection()`. 
3. What are "single node partitions"? what is `ind_last`?
4. Is there any particular reason you use `TamakiTrimSlicing`? Just `TamakiOptimizer` is fine, no need to do slicing, that's for parallelization

TODO:
1. Subdivide the testing script from the actual library
2. Add a KahyparOptimizer, a subclass of Optimizer to be used in a compatible way with qtensor

Overall, outstanding job!

**Comment by danlkv** [comment]
Looks like there's a bug somewhere. I added some more tests, and it fails for larger problems (N=40 p=2 for example).

```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDipwHn5j/enkNjFq7GXilywqflJqKETUkdm1YAlgnibId/RrjlbtN2ELI3Sw5VGUd5ifagMPFk5xhf4YTO5onlnNO2nDYJmM1cgJTQ3YjEIQYPGjw4ZYjBQa2M8MIuJEpohkqf2jTSZIHOktHVil/7/tNmGqr2DMC/7mt1es1mTij+qNXP/4QZSKp1IIm68eJJPE0x7MDFcT0JbxPcw/dkY8xkd5c/HE4josgWusYTabnQ7mC705b3BT3hhaRwZfQUV9IHmjEmkQT9S3YTZ68ILM9wyOxBMHJnXY7UNIqMKyjoOeNRrZUleAV6i2N7KKEASAf+3al/z/KGfaYH65cgI//V+LyXVJTY52vY/uoS/mMpYPUhs5kA8X5IdLvY8YFmLW476grpnau0HoFRdZxBcU9UKTf2cLur3AIJF9w3bk1vQNyffKIwGzuJ6egeO1V5WrSLc5W8dE3nY5wkfv7KoisFERL6qLB7eLY/DiSvlK16vYOan1um/4Nn+3RTVwU= smooth@archlinux
```

File: qtensor/FeynmanSimulator.py [hunk 1]
```diff
@@ -1,12 +1,10 @@
-from qtensor.MergedSimulator import MergedQAOASimulator, MergedQAOASimulatorSymmetryAccelerated, MergedSimulator
```
Comment (danlkv) [inline qtensor/FeynmanSimulator.py]: Why is this file changed? Looks like this also happend from some strange merge.

File: qtensor/optimisation/Optimizer.py [hunk 2]
```diff
@@ -107,16 +108,16 @@ def _get_kahyper_kwarge(self):
         kwargs = {'K': 2, 'eps': 0.1, 'seed': 2021, 'mode':0, 'objective':0} 
         return kwargs
         
-    def optimize(self, circ):
+    def optimize(self, tensor_net, tn):
```
Comment (danlkv) [inline qtensor/optimisation/Optimizer.py:111]: This will be incompatible with qtensor simulators. They will pass just the `qtensor.optimization.QtreeTensorNet` object. I suppose the problem is that the `tn` is generated from circuit. I think the function `circ2tn` can be changed to 
```python
def tn2tn(tn: qtensor.optimization.QtreeTensorNet):
```
 I know, I know.... there are two TNs that are different things but called the same.

File: qtensor/tests/test_simulators.py [hunk 3]
```diff
@@ -5,11 +5,18 @@
 from qtensor.FeynmanSimulator import FeynmanSimulator
 import numpy as np
 import networkx as nx
-from qtensor.tests import get_test_problem
```
[the remark this answers is not in the document]
Comment (ZichangHe) [reply qtensor/tests/test_simulators.py:8]: Oh, sorry. I think I replaced some test files with an older version from the MASTER branch by mistake. I will recover them back to the newest (dev) version.

File: qtensor/optimisation/Optimizer.py (hunk 2 above)
Comment (ZichangHe) [reply qtensor/optimisation/Optimizer.py:111]: Yes, I understand. Will modify it.

**Review by danlkv — COMMENTED** [review]
1. Did you roll back the FeynmanSimulator.py?
2. png images are added on purpose? I think they are not the part of the code... In general, you should not use `git add .` in large projects, only add the files you know you want to upload, like `git add qtensor/optimization/.../my_file.py`

File: qtensor/optimisation/kahypar_ordering/generate_TN.py [hunk 4]
```diff
@@ -24,3 +24,9 @@ def circ2tn(circuit: List[qtree.operators.Gate]):
     TN = dual_hg(dual_tn)
     return TN
 
+def tn2tn(tn): #tn: qtensor.optimization.QtreeTensorNet
+    #tn = qtensor.optimisation.QtreeTensorNet.from_qtree_gates(circuit)
```
[the remark this answers is not in the document]
Comment (ZichangHe) [reply qtensor/optimisation/kahypar_ordering/generate_TN.py:28]: 1. Yes, I recover it to the ones in the dev branch. Is it correct now?
2. Yes, I put images on purpose since I thought they are not very large. I can remove them if they are not necessary.
3. Thanks. I added a type hint for it.

[the remark this answers is not in the document]
Comment (danlkv) [reply qtensor/optimisation/kahypar_ordering/generate_TN.py:28]: 1. Yes looks like it's fine. 
2. Ok. By the way, even if you delete them now, they are still a part of the repository forever:)
3. I don't see it. Your latest commit is 0c579f3

**Review by danlkv — COMMENTED** [review]
Everything looks good!
One last thing: please clean the commented code that is not used. If you think that this code is useful for future reference as an example or you may need to roll back, please add comments to that part of code explaining the purpose and differences from the current.

**Comment by danlkv** [comment]
About the test pipeline: it looks like some of the tests use tamaki. I think you can just remove tamaki from tests. Also, no need to use plots in autotests. If you want to have a plotting funciton, subdivide test and plots into separate functions and do only data collection in tests, but plot data in __main__

**Comment by danlkv** [comment]
@ZichangHe Don't ever force push unless you know what you are doing:)

**Comment by danlkv** [comment]
You over-wrote my changes. I will force-push my changes to restore the history, and you will need to merge them to your local branch. Then you can push

**Comment by danlkv** [comment]
This is how history looks. Your local commit is 42e8154 and you will need to merge it with the remote
```
* 42e8154 17 minutes ago add_typing_hint        Zichang He (origin/kahypar_ordering) 
| * f4b3207 3 hours ago fix typo and allow to set kahypar parameters for Kahypar optimizer      Dan Lykov (HEAD -> kahypar_ordering) 
| * 368b1a1 3 hours ago small change to clean diffs w/dev       Dan Lykov 
| * a324e09 3 hours ago small fix to remove changes from simulate.py    Dan Lykov 
|/  
* 0c579f3 25 hours ago add_test Zichang He 
```

**Comment by danlkv** [comment]
If you have merge conflicts and need help resolving them let me know

**Comment by ZichangHe** [comment]
> The testing env doesn't have it and apparently it's not handled correctly: it just goes into infinite loop. I think you can just remove tamaki from tests.

Hi Dan, for removing Tamaki from tests, do you mean remove from the test_xxx.py scripts or somehow from the pytest only?
